client/client.go

```diff
@@ -724,6 +751,89 @@ func newWebsocketDialerTLS(conf Config) (*websocket.Dialer, error) {
 	return &dialer, nil
 }
 
+var ProxyURLFromHostPortGetter = func(addr string) (*url.URL, error) {
+	u, err := url.Parse("https://" + addr)
+	if err != nil {
+		return u, err
+	}
+	return http.ProxyFromEnvironment(&http.Request{URL: u})
+}
+
+func getHostPort(u *url.URL) (hostPort string) {
+	hostPort = u.Host
+	if i := strings.LastIndex(u.Host, ":"); i > strings.LastIndex(u.Host, "]") {
+		return u.Host
+	} else {
+		switch u.Scheme {
+		case "https":
+			hostPort += ":443"
+		default:
+			hostPort += ":80"
+		}
+	}
+	return hostPort
+}
+
+func dialProxy(network string, addr string, proxyURL *url.URL) (net.Conn, error) {
+	var (
+		resp *http.Response
+		err  error
+	)
+	hostPort := getHostPort(proxyURL)
+	conn, err := net.Dial(network, hostPort)
+	if err != nil {
+		return nil, err
+	}
+
+	connectHeader := make(http.Header)
+	if user := proxyURL.User; user != nil {
+		proxyUser := user.Username()
+		if proxyPassword, passwordSet := user.Password(); passwordSet {
+			credential := base64.StdEncoding.EncodeToString([]byte(proxyUser + ":" + proxyPassword))
+			connectHeader.Set("Proxy-Authorization", "Basic "+credential)
+		}
+	}
+
+	connectReq := &http.Request{
+		Method: http.MethodConnect,
+		URL:    &url.URL{Opaque: addr},
+		Host:   addr,
+		Header: connectHeader,
+	}
+	connectCtx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+	defer cancel()
+
+	didReadResponse := make(chan struct{})
+
+	go func() {
+		defer close(didReadResponse)
+		if err := connectReq.Write(conn); err != nil {
+			return
+		}
+		br := bufio.NewReader(conn)
+		resp, err = http.ReadResponse(br, connectReq)
+	}()
+	select {
+	case <-connectCtx.Done():
+		conn.Close()
+		<-didReadResponse
+		return nil, connectCtx.Err()
+	case <-didReadResponse:
+	}
```

Comment:
I changed this so that it will timeout if no response was received after a minute. @lluiscampos could you take a second look at this?

Reply:
I probably should send down the `ctx` from `NetDialTLSContext` and `DialTLSContext` so that we can spawn `connectCtx` from `ctx`.